### PR TITLE
UPSTREAM: 36071: Add --local & --dry-run flags to `oc set` sub-commands

### PR DESCRIPTION
**Pull request by juanvallejo**
UPSTREAM: https://github.com/kubernetes/kubernetes/pull/36071
Related Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1390139

Related PR (adds `--local` and `--dr-run` flags to `oc set resources`): https://github.com/openshift/origin/pull/11694

Partially addresses: https://github.com/openshift/origin/issues/7899

some `oc set ...` sub-commands do not implement `--dry-run` or `--local`
options. This patch adds these missing options to all sub-commands and
ensures that backwards compatibility remains for commands that treat the
--output option as a ``dry-run flag.

##### Example
```
$ oc set image dc dctest dctest-1=default/testimage:latest --dry-run
NAME      REVISION   DESIRED   CURRENT   TRIGGERED BY
dctest    1          1         1         config

$ oc set image dc dctest dctest-1=default/testimage:latest -o name
deploymentconfig/dctest

$ oc set build-secret mysecret --push --all -o
custom-columns=NAME:.metadata.name
NAME
gitauthtest
repo-base

$ oc set build-secret mysecret --push --all --dry-run
NAME          TYPE      FROM      LATEST
gitauthtest   Source    Git       1
NAME        TYPE      FROM      LATEST
repo-base   Docker    Git       8
```

##### Affected commands

- [x] `oc set build-hook`
- [x] `oc set image` (Also fixed upstream in https://github.com/kubernetes/kubernetes/pull/36071)
- [x] `oc set probe`
- [x] `oc set resources` (fixed in https://github.com/openshift/origin/pull/11694)
- [x] `oc set triggers`
- [x] `oc set volume`
- [x] `oc set env`
- [x] `oc set deployment-hook`
- [x] `oc set build-secret`
- [x] `oc set route-backends`

cc @openshift/cli-review

### CODE REVIEW

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/set/set_image.go [hunk 1]
```diff
@@ -134,13 +134,11 @@ func (o *ImageOptions) Complete(f *cmdutil.Factory, cmd *cobra.Command, args []s
 		ContinueOnError().
 		NamespaceParam(cmdNamespace).DefaultNamespace().
 		FilenameParam(enforceNamespace, o.Recursive, o.Filenames...).
+		SelectorParam(o.Selector).
+		ResourceTypeOrNameArgs(o.All, o.Resources...).
+		Latest().
```
Comment (liggitt) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/set/set_image.go]: Latest() does a server fetch, right? I thought the point of Local was to avoid that

**Comment by juanvallejo** [comment]
@liggitt Thanks for the feedback, review comment addressed

**Comment by ncdc** [comment]
Needs an actual upstream PR. Can this wait to come in via rebase? cc @pweil- @mfojtik

**Comment by mfojtik** [comment]
@ncdc i'm fine waiting for the rebase

**Comment by juanvallejo** [comment]
[test]

**Comment by juanvallejo** [comment]
conformance check flaked on https://github.com/openshift/origin/issues/10988 re[test]

**Comment by juanvallejo** [comment]
make verify flaked on https://github.com/openshift/origin/issues/11672 re[test]

**Comment by fabianofranz** [comment]
The last failure complained about outdated docs, did you `hack/update-generated-docs.sh`recently?

**Comment by juanvallejo** [comment]
Yeah, I updated docs right after that failure, however it has been failing with `tput: No value for $TERM and no -T specified` since then. (Although locally it fails with:
```
===== Verifying Generated Bootstrap Bindata =====
Generating bootstrap bindata...
FAILURE: Generation of fresh bindata failed:
/home/jvallejo/Documents/go/src/github.com/openshift/origin/hack/update-generated-bootstrap-bindata.sh: line 18: : command not found
[ERROR] PID 16909: hack/update-generated-bootstrap-bindata.sh:18: `"$(os::util::find-go-binary go-bindata)" -nocompress -nometadata -prefix "bootstrap" -pkg "bootstrap" -o "${OUTPUT_PARENT}/pkg/bootstrap/bindata.go" -ignore "README.md" -ignore ".*\.go$" -ignore application-template.json ${EXAMPLES}/image-streams/... ${EXAMPLES}/db-templates/... ${EXAMPLES}/jenkins ${EXAMPLES}/jenkins/pipeline ${EXAMPLES}/quickstarts/... ${EXAMPLES}/logging/... pkg/image/admission/imagepolicy/api/v1/...` exited with status 127.
```
)

**Comment by juanvallejo** [comment]
@stevekuznetsov I keep getting `tput: No value for $TERM and no -T specified` on the Travis test, and 
this locally when running `GO_PERMISSIVE=y make verify`
```
===== Verifying Generated Bootstrap Bindata =====
Generating bootstrap bindata...
FAILURE: Generation of fresh bindata failed:
/home/jvallejo/Documents/go/src/github.com/openshift/origin/hack/update-generated-bootstrap-bindata.sh: line 18: : command not found
[ERROR] PID 16909: hack/update-generated-bootstrap-bindata.sh:18: `"$(os::util::find-go-binary go-bindata)" -nocompress -nometadata -prefix "bootstrap" -pkg "bootstrap" -o "${OUTPUT_PARENT}/pkg/bootstrap/bindata.go" -ignore "README.md" -ignore ".*\.go$" -ignore application-template.json ${EXAMPLES}/image-streams/... ${EXAMPLES}/db-templates/... ${EXAMPLES}/jenkins ${EXAMPLES}/jenkins/pipeline ${EXAMPLES}/quickstarts/... ${EXAMPLES}/logging/... pkg/image/admission/imagepolicy/api/v1/...` exited with status 127.
```

was `hack/update-generated-bootstrap-bindata.sh` being run before?

**Comment by stevekuznetsov** [comment]
@juanvallejo travis bug is fixed with #11877

**Comment by fabianofranz** [comment]
flaked on https://github.com/openshift/origin/issues/10228 and https://github.com/openshift/origin/issues/11771 and https://github.com/openshift/origin/issues/11876 re[test]

**Comment by ncdc** [comment]
Closing as we'll get this with the kube 1.5 rebase.

**Comment by juanvallejo** [comment]
Reopening as the kube rebase won't give us fixes for the rest of the `oc set...` subcommands

**Comment by ncdc** [comment]
Sorry, I didn't realize you had made origin changes.

**Comment by ncdc** [comment]
And FYI no need to chase getting the tests to pass now. It won't be ready for merging until sometime in December at the earliest.

**Comment by juanvallejo** [comment]
@ncdc 
> And FYI no need to chase getting the tests to pass now. It won't be ready for merging until sometime in December at the earliest.

Sounds good, just want to make sure tests are flakes / figure out which change is causing conformance to keep failing

**Comment by fabianofranz** [comment]
@juanvallejo complaining about man pages, can you update?

**Comment by juanvallejo** [comment]
conformance flaked on https://github.com/openshift/origin/issues/11887 re[test]

**Comment by fabianofranz** [comment]
Ok, LGTM. Need to remove WIP and squash the local commits. @liggitt do we want this in 1.4?

**Comment by liggitt** [comment]
not high enough severity

**Comment by ncdc** [comment]
We've already discussed that this is not coming in 1.4: https://github.com/openshift/origin/pull/11733#issuecomment-258091738 and https://github.com/openshift/origin/pull/11733#issuecomment-259974115. Also please see the milestone is 1.5.0.

**Comment by fabianofranz** [comment]
[merge]

**Comment by ncdc** [comment]
@fabianofranz I deleted your merge tag. This has to wait until after the kube 1.5 rebase is in.

**Comment by fabianofranz** [comment]
:+1:

**Comment by ncdc** [comment]
Also how important is this? kube pull 36071 hasn't merged yet, so it's not in kube 1.5.

**Comment by juanvallejo** [comment]
integration test flaked on https://github.com/openshift/origin/issues/8571
 re[test]

**Comment by juanvallejo** [comment]
integration flaked on https://github.com/openshift/origin/issues/8571
 re[test]

**Comment by juanvallejo** [comment]
conformance check flaked on https://github.com/openshift/origin/issues/8571
 re[test]

**Comment by juanvallejo** [comment]
@fabianofranz Upstream PR meged: https://github.com/kubernetes/kubernetes/pull/36071

**Comment by juanvallejo** [comment]
integration flaked on https://github.com/openshift/origin/issues/8571
 re[test]

**Comment by ncdc** [comment]
Kube pr 36071 is in master but not 1.5. How critical is this for Origin 1.5 or can it wait until 1.6? @fabianofranz @mfojtik @smarterclayton @pweil-

**Comment by smarterclayton** [comment]
Not critical in my mind

**Comment by fabianofranz** [comment]
Can wait, I'm tagging this `post-rebase` and milestone `1.6.0`.

**Comment by fabianofranz** [comment]
@juanvallejo need to squash the local and upstream commits.

**Comment by ncdc** [comment]
Since this is after the 1.6 rebase, there's no need to squash or do anything else with this now. Once Kube 1.6 comes in, the UPSTREAM commits will come out of this PR.

**Comment by fabianofranz** [comment]
@juanvallejo legit failure?

```
FAILURE after 0.920s: test/cmd/basicresources.sh:208: executing 'oc set probe dc test-deployment-config --liveness --open-tcp=8080' expecting success and text 'updated': the command returned the wrong error code; the output content test failed
```

**Comment by juanvallejo** [comment]
@fabianofranz hm, I can't seem to reproduce this locally, rebasing with latest master and once more

**Comment by juanvallejo** [comment]
@fabianofranz meant to force-push after updating `set-image.sh`, `set-liveness-probe.sh`, but deleted branch on accident instead. Apparently I cannot restore the branch once I have force pushed after its deletion. 

Re-opened here: https://github.com/openshift/origin/pull/14123

**Pull request closed without merging**